// inngest/components/DeferredFunction.cjs
const require_consts = require('../helpers/consts.cjs');
const require_errors = require('../helpers/errors.cjs');
const require_types = require('../helpers/types.cjs');
const require_marker = require('../helpers/marker.cjs');
const require_InngestFunction = require('./InngestFunction.cjs');

//#region src/components/DeferredFunction.ts
const idDenyRegex = /['\\\n\r]/;
/**
* Validates the reserved experiment-ref input key into an `ExperimentRef`, or
* `undefined` when it's absent or malformed.
*/
function parseExperimentRef(value) {
	if (require_types.isRecord(value) && typeof value.experimentName === "string" && typeof value.variant === "string") return {
		experimentName: value.experimentName,
		variant: value.variant
	};
}
/**
* Strips our `_inngest` metadata off an event's `data` in place and returns the
* parent routing metadata extracted from it.
*/
function stripInngestMetadata(event) {
	const { _inngest, [require_consts.deferExperimentKey]: experiment, ...input } = event.data ?? {};
	if (!require_types.isRecord(_inngest)) throw new require_errors.UnreachableError("deferred event is missing _inngest metadata");
	const { parent_fn_slug, parent_run_id } = _inngest;
	if (typeof parent_fn_slug !== "string") throw new require_errors.UnreachableError("deferred event _inngest metadata is missing parent_fn_slug");
	if (typeof parent_run_id !== "string") throw new require_errors.UnreachableError("deferred event _inngest metadata is missing parent_run_id");
	event.data = input;
	return {
		fnSlug: parent_fn_slug,
		runId: parent_run_id,
		experiment: parseExperimentRef(experiment)
	};
}
/**
* EXPERIMENTAL: This API is not yet stable and may change in the future without
* a major version bump.
*
* A defer (companion) function created via `createDefer(...)`. Real
* `InngestFunction` at runtime, but with the trigger pinned to
* `inngest/deferred.schedule` (see `getConfigTriggers`), `triggers` and
* `onFailure` disallowed, and the schema carried as a typed instance
* property so callers of `defer(id, { function, data })` can extract it.
*
* Identify a defer function at runtime via `isDeferredFunction(value)` from
* `helpers/marker.ts`. Prefer that over `instanceof`, which fails across
* duplicate SDK copies in the same process.
*
* @public
*/
var DeferredFunction = class extends require_InngestFunction.InngestFunction {
	schema;
	[require_marker.markerKey] = { kind: "DeferredFunction" };
	constructor(client, opts, handler, schema) {
		if (idDenyRegex.test(opts.id)) throw new Error(`invalid id "${opts.id}"; must match ${idDenyRegex.source}`);
		super(client, {
			...opts,
			triggers: []
		}, handler);
		this.schema = schema;
	}
	getConfigTriggers(fnId) {
		return [{
			event: require_consts.internalEvents.DeferredSchedule,
			expression: `event.data._inngest.fn_slug == '${fnId}'`
		}];
	}
	/**
	* Installs a `transformCtx` hook that runs before middleware and user code:
	* - Strips our `_inngest` metadata off each event's `data`, ensuring it
	*   matches what the user expects.
	* - Collects the parent function's slug and run ID from that metadata into
	*   `ctx.parents`, so the handler can correlate each event to its parent.
	*/
	createExecution(opts) {
		return super.createExecution({
			...opts,
			partialOptions: {
				...opts.partialOptions,
				transformCtx: (ctx) => {
					const parents = ctx.events.map(stripInngestMetadata);
					stripInngestMetadata(ctx.event);
					return {
						...ctx,
						parents
					};
				}
			}
		});
	}
};
/**
* EXPERIMENTAL: This API is not yet stable and may change in the future without
* a major version bump.
*
* Create a typed defer function. One `createDefer` call = one Inngest
* function. Returns a `DeferredFunction<TSchema>` so callers of `defer(id,
* { function, data })` get the data type inferred from the schema.
*
* Mirrors `inngest.createFunction(opts, handler)`, with three differences:
* the client is the first positional arg, `triggers` is not accepted (the
* SDK emits an implicit `inngest/deferred.schedule` trigger), and `schema`
* describes the payload that callers will send via `defer(id, { function,
* data })`.
*
* Pass the result to `serve()` alongside regular functions so the SDK
* registers it.
*/
function createDefer(client, options, handler) {
	const { schema, ...rest } = options;
	return new DeferredFunction(client, rest, handler, schema);
}

//#endregion
exports.DeferredFunction = DeferredFunction;
exports.createDefer = createDefer;
