// inngest/components/InngestScore.js
import { isFiniteNumber, isRecord } from "../helpers/types.js";
import { Middleware } from "./middleware/middleware.js";
import { performOp } from "./InngestMetadata.js";

//#region src/components/InngestScore.ts
const scoreKind = "inngest.score";
const experimentKind = "inngest.experiment";
const maxScoreNameByteLength = 128;
const scoreSymbol = Symbol.for("inngest.step.score");
function validateIdField({ value, field, required }) {
	if (!required && value === void 0) return;
	if (!(typeof value === "string" && value.trim().length > 0)) throw new Error(`${field} must be a non-empty string`);
}
function validateScoreFields(options, requiredTargetIds) {
	if (!isRecord(options)) throw new Error("score options must be an object");
	for (const field of ["runId", "stepId"]) validateIdField({
		value: options[field],
		field,
		required: requiredTargetIds.includes(field)
	});
	if (typeof options.name !== "string" || options.name.trim().length === 0) throw new Error("score name must be a non-empty string");
	if (/[\x00-\x1f\x7f']/.test(options.name)) throw new Error("score name must not contain control characters or single quotes");
	const nameByteLength = new TextEncoder().encode(options.name).length;
	if (nameByteLength > maxScoreNameByteLength) throw new Error(`score name must be ${maxScoreNameByteLength} bytes or fewer in UTF-8 (got ${nameByteLength})`);
	if (typeof options.value !== "boolean" && !isFiniteNumber(options.value)) throw new Error("score value must be a finite number or boolean");
}
function validateSendScoreOptions(options) {
	validateScoreFields(options, []);
}
function validateStepScoreOptions(options) {
	validateScoreFields(options, []);
}
async function sendScore(client, options) {
	validateSendScoreOptions(options);
	await performOp(client, {
		runId: options.runId,
		stepId: options.stepId
	}, { [options.name]: { value: options.value } }, `${scoreKind}`, "merge");
}
async function sendStepScore(client, options) {
	validateStepScoreOptions(options);
	await performOp(client, {
		runId: options.stepId === void 0 ? options.runId ?? null : options.runId,
		stepId: options.stepId
	}, { [options.name]: { value: options.value } }, `${scoreKind}`, "merge");
}
function validateExperimentRef(experiment) {
	if (!isRecord(experiment)) throw new Error("experiment must be an object");
	for (const field of ["experimentName", "variant"]) if (typeof experiment[field] !== "string" || experiment[field].trim().length === 0) throw new Error(`experiment.${field} must be a non-empty string`);
}
async function sendScoreExperiment(client, options) {
	validateSendScoreOptions(options);
	validateExperimentRef(options.experiment);
	const target = {
		runId: options.runId,
		stepId: options.stepId
	};
	await performOp(client, target, {
		name: options.experiment.experimentName,
		variant: options.experiment.variant
	}, experimentKind, "merge");
	await performOp(client, target, { [options.name]: { value: options.value } }, scoreKind, "merge");
}
const scoreMiddleware = () => {
	class ScoreMiddleware extends Middleware.BaseMiddleware {
		id = "inngest:score";
		static onRegister({ client }) {
			client["experimentalScoreEnabled"] = true;
		}
		transformFunctionInput(arg) {
			return {
				...arg,
				ctx: {
					...arg.ctx,
					step: {
						...arg.ctx.step,
						score: arg.ctx.step[scoreSymbol]
					}
				}
			};
		}
	}
	return ScoreMiddleware;
};

//#endregion
export { scoreMiddleware, scoreSymbol, sendScore, sendScoreExperiment, sendStepScore, validateStepScoreOptions };
